lib/cpp/src/thrift/server/TNonblockingServer.cpp

```diff
@@ -1524,7 +1532,7 @@ void TNonblockingIOThread::breakLoop(bool error) {
   }
 
   // sets a flag so that the loop exits on the next event
-  event_base_loopbreak(eventBase_);
+  event_base_loopexit(eventBase_, 0);
```

Comment:
There are some subtle differences between the two. The first difference is what you already mentioned. The second difference is that, event_base_loopbreak will have no effect when there is no loop running: http://www.wangafu.net/~nickm/libevent-book/Ref3_eventloop.html. In contrast, event_base_loopexit is not `lost` if it is called without having a loop running.

In the TNonblockingServer code, event_base_loopbreak is called without checking if the loop started running.

An alternative fix would be to keep the event_base_loopbreak, but somehow introduce additional synchronization to make sure it is not lost.